frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/IncrementalBuildExecutionDigest.java

```diff
@@ -0,0 +1,214 @@
+    /**
+     * Must be something that would be encoded otherwise we could end up with a set of
+     * parts we don't expect.
+     */
+    static final String SERIALIZATION_SEPARATOR = ";";
+
+    public Long digestVersion;
+
+    @JsonDeserialize(keyUsing = ExecutionCoordinatesDeserializer.class)
+    @JsonSerialize(keyUsing = ExecutionCoordinatesSerializer.class)
+    public Map<ExecutionCoordinates, Execution> executions;
+
+    public IncrementalBuildExecutionDigest() {
+            // for Jackson
+    }
+
+    public IncrementalBuildExecutionDigest(Long digestVersion, Map<ExecutionCoordinates, Execution> executions) {
+        this.digestVersion = digestVersion;
+        this.executions = executions;
+    }
+
+    public static class ExecutionCoordinates {
+        public String goal;
+        public String id;
+        public String lifecyclePhase;
+
+        public ExecutionCoordinates() {
+            // for Jackson
+        }
+
+        public ExecutionCoordinates(String goal, String id, String lifecyclePhase) {
+            this.goal = goal;
+            this.id = id;
+            this.lifecyclePhase = lifecyclePhase;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (!(o instanceof ExecutionCoordinates)) return false;
+            ExecutionCoordinates that = (ExecutionCoordinates) o;
+            return Objects.equals(goal, that.goal) && Objects.equals(id, that.id) && Objects.equals(lifecyclePhase, that.lifecyclePhase);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(goal, id, lifecyclePhase);
+        }
+    }
+
+    public static class Execution {
+        public String arguments;
+        public Map<String, String> environmentVariables;
+        public Set<File> files;
+        public Runtime runtime;
+
+        public Execution() {
+            // for Jackson
+        }
+
+        public Execution(String arguments, Map<String, String> environmentVariables, Set<File> files, Runtime runtime) {
+            this.files = files;
+            this.environmentVariables = environmentVariables;
+            this.arguments = arguments;
+            this.runtime = runtime;
+        }
+
+        public static class File {
+            public String filename;
+            public Integer byteLength;
+            public String hash;
+
+            public File() {
+                // for Jackson
+            }
+
+            public File(String filename, Integer byteLength, String hash) {
+                this.filename = filename;
+                this.byteLength = byteLength;
+                this.hash = hash;
+            }
+
+            @Override
+            public boolean equals(Object o) {
+                if (!(o instanceof File)) return false;
+                File file = (File) o;
+                return Objects.equals(filename, file.filename) && Objects.equals(byteLength, file.byteLength) && Objects.equals(hash, file.hash);
```

Comment:
nit: the bytelength is *most probably* redundant ;)

Reply:
Almost all the time, but it's basically free